### PR TITLE
Command aliases

### LINKED ISSUE #823: Subcommand aliases
Problem
-------

The `command()` and `subcommand()` primitives accept a single name string. Many real-world CLIs expose shorter aliases for frequently used subcommands—for example, `git ci` for `git commit`, `npm i` for `npm install`, or `docker ps` for `docker container ls`.

Currently, the only way to create an alias is to duplicate the entire subcommand parser tree and wrap both in `or()`:

~~~~ typescript
const installCmd = command("install", installParser, { description: "Install a package" });
const iCmd = command("i", installParser, { description: "Install a package", hidden: true });

const cli = command("mytool", object({
  sub: subcommand(or(installCmd, iCmd)),
}));
~~~~

This is verbose, error-prone (the two definitions can drift apart), and the hidden alias still appears as a separate entry in the internal parser tree even though it carries the same help text.

Proposed solution
-----------------

Accept an optional `aliases` field in the command options:

~~~~ typescript
interface CommandOptions {
  readonly description?: Message;
  readonly hidden?: HiddenVisibility;
  // ...existing options...

  /**
   * Additional names that invoke this command.
   * Aliases are functional at runtime but hidden from help output.
   * The first name in the `command()` call is treated as the canonical
   * display name.
   *
   * @since TBD
   */
  readonly aliases?: readonly [string, ...string[]];
}
~~~~

Usage:

~~~~ typescript
const installCmd = command("install", installParser, {
  description: "Install a package",
  aliases: ["i"],
});
~~~~

### Behavior details

 -  At parse time, any alias triggers the same parser as the canonical name.
 -  In help output, only the canonical name is shown (aliases are never rendered in usage lines, option tables, or subcommand lists).
 -  In shell completions, aliases are suggested alongside the canonical name—typing `git i<TAB>` should complete to `git i`.
 -  In error messages (“unknown command”), suggestions include aliases as candidates so that a typo of an alias can still suggest the canonical name.
 -  Aliases must not collide with sibling command names or other aliases within the same parent. This is validated at construction time with a `TypeError`.

### Interaction with `negatableFlag()` and help/version/completion meta options

The project already has a precedent for “display name + hidden aliases” in the help and version option configuration (`RunOptions.help` and `RunOptions.version` support an array where the first element is visible and the rest are hidden aliases). The same pattern applies here—the first name passed to `command()` is the display name, and everything in `aliases` is a hidden functional alias.

Alternatives considered
-----------------------

 -  **`command(["install", "i"], ...)`**: accepting an array as the first argument. Rejected because it makes the common single-name case more verbose and diverges from the existing API shape.
 -  **A separate `alias()` combinator**: `alias(command("install", ...), "i")`. Rejected because it introduces a new combinator for what is essentially a command-level concern, and the help-text semantics (hide aliases) are more naturally expressed inside the command options.

### PR DESCRIPTION
**Pull request by dahlia**
Summary
-------

This PR adds command aliases to `command()`, so a parser can accept short or legacy command names while keeping one canonical command path and display name. Aliases are accepted during parsing, included in shell completion and typo suggestions, and kept out of usage, help, and documentation output.

Fixes #823.

What changed
------------

 -  Added `aliases` to `CommandOptions` in *@optique/core*.
 -  Preserved canonical command paths even when the input uses an alias.
 -  Included aliases in completion and typo suggestions.
 -  Rejected command name and alias collisions among active sibling alternatives in `or()`, `longestMatch()`, `object()`, and `merge()`.
 -  Kept ordered compositions such as `tuple()` and `seq()` positional, and documented why duplicate aliases are allowed there.
 -  Updated *docs/concepts/primitives.md*, *docs/concepts/completion.md*, and *CHANGES.md*.

Validation
----------

 -  `deno test packages/core/src/facade.test.ts packages/core/src/primitives.test.ts packages/core/src/constructs.test.ts packages/core/src/usage.test.ts`
 -  `mise check`
 -  `cd docs && pnpm build`
 -  `mise test`
 -  `deno fmt --check`
 -  `git diff --check`

### CODE REVIEW

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/suggestion.ts [hunk 1]
```diff
@@ -254,6 +258,70 @@ export function createSuggestionMessage(
...
+  for (const suggestion of suggestions) {
+    const targets = commandAliasTargets.get(suggestion) ?? [suggestion];
+    for (const target of targets) {
+      if (seen.has(target)) continue;
+      seen.add(target);
+      expanded.push(target);
+    }
+  }
+  return expanded;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:288]: Addressed in b12355bf: expandCommandAliasSuggestions now returns immediately when there are no suggestions to expand.

File: packages/core/src/constructs.test.ts [hunk 2]
```diff
@@ -22351,6 +22394,58 @@ describe("seq", () => {
...
+          object({
+            install: command("install", object({}), { aliases: ["i"] }),
+          }),
+          object({
+            inspect: command("inspect", object({}), { aliases: ["i"] }),
+          }),
+        ),
+      TypeError,
+    );
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.test.ts:22610]: Addressed in 25d6be87: the alias collision tests now assert the diagnostic message as well as the TypeError.

File: packages/core/src/facade.test.ts [hunk 3]
```diff
@@ -4446,10 +4446,13 @@ describe("Subcommand help edge cases (Issue #26 comprehensive coverage)", () =>
       assert.ok(!suggestions.includes("--version"));
     });
 
-    it("should not treat nested duplicate command branch options as current value slots", () => {
-      const parser = or(
-        command("tool", command("deploy", option("--target", string()))),
-        command("tool", command("status", object({}))),
+    it("should not treat nested command options as current value slots", () => {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts]: Addressed in 25d6be87: the two nested command completion tests now have distinct names for the unmatched and unselected branch cases.

File: packages/core/src/primitives.ts [hunk 4]
```diff
@@ -2895,10 +2907,28 @@ function appendCommandPath(
   };
 }
 
+function getCommandNames(
+  name: string,
+  options: CommandOptions,
+): readonly string[] {
+  return options.aliases == null ? [name] : [name, ...options.aliases];
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/primitives.ts:2962]: Addressed in 25d6be87: command aliases are now validated as a non-empty string array before they are spread.

File: packages/core/src/constructs.ts [hunk 5]
```diff
@@ -3989,6 +4047,13 @@ export interface LongestMatchOptions {
   errors?: LongestMatchErrorOptions;
 }
 
+const allowDuplicateLeadingCommandNamesKey =
+  "__optiqueAllowDuplicateLeadingCommandNames";
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in 7ca6f47b: the key now lives in an internal module and constructs imports it from there, so it is shared without becoming part of the public constructs API.

File: packages/core/src/facade.ts [hunk 6]
```diff
@@ -78,6 +78,18 @@ import type {
   SourceContextRequest,
 } from "./context.ts";
 
+const allowDuplicateLeadingCommandNamesKey =
+  "__optiqueAllowDuplicateLeadingCommandNames";
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Addressed in 7ca6f47b: facade now imports the shared internal key instead of duplicating the string locally.

File: packages/core/src/internal/command-alias.ts [hunk 7]
```diff
@@ -0,0 +1,2 @@
+export const allowDuplicateLeadingCommandNamesKey =
+  "__optiqueAllowDuplicateLeadingCommandNames";
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/internal/command-alias.ts]: Addressed in 581b1238.

**Comment by dahlia** [comment]
I skipped the outside-diff hidden-alias suggestion for `createUnexpectedInputErrorWithScopedSuggestions()`. Visible command aliases are already included by `collectLeadingCandidates()` through the command term aliases, while `hidden: true` commands and options are intentionally excluded from typo suggestions. I verified the suggested hidden-inclusive matcher change against the existing hidden suggestion tests, and it would regress `parser-error-suggestions.test.ts`.

**Comment by dahlia** [comment]
Addressed the outside-diff `facade.ts` meta-command composition suggestions in b851ffb. The internal longestMatch() bypass no longer uses `as never`, and both meta-command composition paths now use the same variadic helper.

**Comment by dahlia** [comment]
Addressed the outside-diff shared-buffer reachability issue in 86b1834. `object()` and `merge()` now skip command alias collisions that are unreachable behind a higher-priority catch-all parser, while `or()` and `longestMatch()` still reject active sibling collisions.

**Comment by dahlia** [comment]
Skipped the outside-diff suggestion to base duplicate command alias validation directly on `parser.leadingNames`. `leadingNames` is generic parser metadata and can describe options, literals, and custom token parsers as well as commands, so it cannot distinguish command aliases or their canonical display names. The shared-buffer reachability part of that feedback was valid and is addressed in 86b1834.

**Comment by dahlia** [comment]
Addressed the parser-metadata duplicate command check in 5e49975a8d97ac28ada5390e4c69c7984b6aeba5. The check now reads parser leadingNames, priority, and acceptingAnyToken directly, keeps or() and longestMatch() as all-active alternatives, and applies reachability blocking only to shared-buffer compositions.

**Comment by dahlia** [comment]
Addressed the meta-command alias fan-out in b204559447a1b8430bbceb487e31c3ee5ec9c321. Help, version, and completion meta commands now use one canonical command parser with aliases attached to it, so alias invocations resolve documentation and help usage through the visible command name.

File: packages/core/src/suggestion.ts [hunk 8]
```diff
@@ -254,6 +258,72 @@ export function createSuggestionMessage(
...
+        term.type === "sequence"
+      ) {
+        traverse(term.terms);
+        continue;
+      }
+      if (term.type === "exclusive") {
+        for (const branch of term.terms) traverse(branch);
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts]: Addressed in c04af7a6; alias expansion now stops at blocking usage terms so nested command aliases cannot preempt active leading aliases.

File: packages/core/src/suggestion.ts [hunk 9]
```diff
@@ -254,6 +258,72 @@ export function createSuggestionMessage(
...
+
+  traverse(usage);
+  return targets;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:357]: Addressed in c04af7a6; alias expansion now follows leading-candidate reachability and ignores aliases behind command boundaries.

File: packages/core/src/constructs.ts [hunk 10]
```diff
@@ -1030,6 +1056,87 @@ function checkDuplicateOptionNames(
...
+  }
+}
+
+function isNonCommandOptionLeadingName(
+  parser: Parser<Mode, unknown, unknown>,
+  name: string,
+): boolean {
+  if (!/^(--|[-/+])/.test(name)) return false;
+  return !extractCommandNames(parser.usage, true).has(name);
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in ef61c4ec; the duplicate-command checks now reuse each parser command-name set while scanning leading names.

[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in ef61c4ec; command names are extracted once per parser before scanning leading names.

File: packages/core/src/constructs.ts [hunk 11]
```diff
@@ -1030,6 +1056,81 @@ function checkDuplicateOptionNames(
...
+
+function checkDuplicateReachableLeadingCommandNames(
+  parserSources: ReadonlyArray<
+    readonly [string | symbol, Parser<Mode, unknown, unknown>]
+  >,
+): void {
+  const commandNameSources = new Map<string, (string | symbol)[]>();
+  const sortedSources = parserSources.toSorted(([, parserA], [, parserB]) =>
+    parserB.priority - parserA.priority
+  );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in b70ee586; the collision check now copies parserSources before sorting, so it no longer depends on Array.prototype.toSorted at runtime.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/primitives.ts [hunk 12]
```diff
@@ -3078,17 +3134,24 @@ export function command<M extends Mode, T, TState>(
       // Handle different states
       if (state === undefined) {
         // Check if buffer starts with our command name
-        if (context.buffer.length < 1 || context.buffer[0] !== name) {
+        if (
+          context.buffer.length < 1 ||
+          !commandNames.includes(context.buffer[0])
+        ) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/primitives.ts:3172]: Addressed in 9b1e75d2. getDocPage now matches command aliases while replaying help targets, so alias-based documentation resolves to the canonical command page.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/suggestion.ts [hunk 13]
```diff
@@ -254,6 +258,93 @@ export function createSuggestionMessage(
...
+
+      if (term.type === "exclusive") {
+        let allSkippable = true;
+        for (const branch of term.terms) {
+          const branchSkippable = traverse(branch);
+          allSkippable = allSkippable && branchSkippable;
+        }
+        if (allSkippable) continue;
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:348]: Addressed in 5f3f9a26. Exclusive usage now continues when any branch is skippable, while still collecting aliases from active branches.

File: packages/core/src/suggestion.ts [hunk 14]
```diff
@@ -254,6 +258,93 @@ export function createSuggestionMessage(
...
+      if (term.type === "optional") {
+        traverse(term.terms);
+        continue;
+      }
+
+      if (term.type === "multiple") {
+        traverse(term.terms);
+        if (term.min === 0) continue;
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:333]: Addressed in 5f3f9a26. Required repeated terms now keep traversing when their inner usage is skippable.

File: packages/core/src/constructs.ts [hunk 15]
```diff
@@ -1030,6 +1056,81 @@ function checkDuplicateOptionNames(
...
+  parserSources: ReadonlyArray<
+    readonly [string | symbol, Parser<Mode, unknown, unknown>]
+  >,
+): void {
+  const commandNameSources = new Map<string, (string | symbol)[]>();
+
+  for (const [source, parser] of parserSources) {
+    const commandNames = extractCommandNames(parser.usage, true);
+    for (const name of parser.leadingNames) {
+      if (/^(--|[-/+])/.test(name) && !commandNames.has(name)) continue;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in 5f3f9a26. Duplicate command checks now consider only leading names that are also exposed as command names in usage.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/suggestion.ts [hunk 16]
```diff
@@ -254,6 +258,93 @@ export function createSuggestionMessage(
...
+  usage: Usage,
+): Map<string, readonly string[]> {
+  const targets = new Map<string, readonly string[]>();
+
+  function traverse(terms: Usage): boolean {
+    if (!terms || !Array.isArray(terms)) return true;
+    for (const term of terms) {
+      if (term.type === "option" || term.type === "argument") {
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts]: Addressed in 2b02d452. Alias target collection now skips option terms, so later command aliases can still expand to their canonical command names.

File: packages/core/src/constructs.ts [hunk 17]
```diff
@@ -670,9 +672,13 @@ function createUnexpectedInputErrorWithScopedSuggestions(
     candidates,
     DEFAULT_FIND_SIMILAR_OPTIONS,
   );
+  const displaySuggestions = expandCommandAliasSuggestions(
+    parsers.flatMap((parser) => parser.usage),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/constructs.ts]: Addressed in 2b02d452. Scoped suggestions now pass sibling parser usages as exclusive branches, preserving aliases from later sibling commands.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/internal/parser.ts [hunk 18]
```diff
@@ -1496,7 +1498,50 @@ function findCommandInCurrentUsageTerm(
   return null;
 }
 
+function commandTermMatches(term: UsageTerm, commandName: string): boolean {
+  return term.type === "command" &&
+    (term.name === commandName || term.aliases?.includes(commandName) === true);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/internal/parser.ts]: Addressed in 707674a7. Alias-based getDocPage navigation now applies command usageLine overrides through the same alias-aware command matcher used for usage lookup.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/internal/parser.ts [hunk 19]
```diff
@@ -1496,7 +1498,50 @@ function findCommandInCurrentUsageTerm(
...
+  usage: Usage,
+  commandName: string,
+  names: Set<string>,
+): void {
+  for (const term of usage) {
+    if (term.type === "command") {
+      if (term.name === commandName) {
+        names.add(term.name);
+        for (const alias of term.aliases ?? []) names.add(alias);
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/internal/parser.ts]: Addressed in 6df426f2. collectCommandInputNames() now reuses commandTermMatches(), so the helper stays consistent with alias-aware command matching.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 20]
```diff
@@ -616,28 +646,19 @@ function createHelpParser(
...
-      );
-    }
-    helpCommand = commandParsers.length === 1
-      ? commandParsers[0]
-      : longestMatch(...commandParsers);
+    const aliases = getMetaCommandAliases(names);
+    helpCommand = command(names[0], innerParser, {
+      description: message`Show help information.`,
+      ...(aliases != null ? { aliases } : {}),
+      hidden: commandConfig.hidden,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:663]: Addressed in 9cc315b7. Meta command aliases now use an internal hidden-alias path, so they remain accepted while completion and typo suggestions display only canonical command names.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/suggestion.ts [hunk 21]
```diff
@@ -254,6 +258,102 @@ export function createSuggestionMessage(
...
+        for (const branch of term.terms) {
+          const branchSkippable = traverse(branch);
+          anySkippable = anySkippable || branchSkippable;
+        }
+        if (anySkippable) continue;
+        return false;
+      }
+    }
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:353]: Addressed in 9f4a4601. collectCommandAliasTargets() now treats unhandled usage terms as non-skippable, so aliases behind literal or future blocking terms are not expanded as reachable suggestions.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/internal/parser.ts [hunk 22]
```diff
@@ -1496,7 +1498,53 @@ function findCommandInCurrentUsageTerm(
   return null;
 }
 
+function commandTermMatches(term: UsageTerm, commandName: string): boolean {
+  return term.type === "command" &&
+    (term.name === commandName ||
+      term.aliases?.includes(commandName) === true ||
+      term.hiddenAliases?.includes(commandName) === true);
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/internal/parser.ts]: Addressed in b63ee5e9. commandTermMatches() now accepts nullish terms and uses optional chaining before checking command names or aliases.

**Comment by dahlia** [comment]
@codex review
